### ISSUE TITLE
User Container killed by docker due to it exceeded the request memory

### ISSUE THREAD

**Issue by deepmeng** [microsoft/pai#5464]
Hi, I met the error as follows. Do you have any ideas?

**Organization Name**:

**Short summary about the issue/question**:

Exit Code: -1200
Exit Type: USER_FAILURE
Exit Reason: User Container killed by docker due to it exceeded the request memory
Pod failed: PodPattern matched: containers: - name: app reason: OOMKilled message: "nel_size=(1,), stride=(1,), bias=False)\n (1): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n (2): ReLU()\n (3): Dropout(p=0.3, inplace=False)\n (4): Conv1d(256, 256, kernel_size=(1,), stride=(1,), bias=False)\n (5): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n (6): ReLU()\n (7): Conv1d(256, 7, kernel_size=(1,), stride=(1,))\n )\n )\n )\n)\n2021-04-30 07:10:43,386 INFO **********************Start training waymo_models/pv_rcnn(default)********************** /usr/local/pai/runtime: line 30: 7 Killed ${RUNTIME_SCRIPT_DIR}/runtime_entry.sh\n" code: 137

Exit Solutions:
  Increase per task memory request
  Decrease per task memory usage by such as increasing task number
**Brief what process you are following**:

**How to reproduce it**:

**OpenPAI Environment**:
- OpenPAI version: v1.4.1
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release): ubuntu 
- Kernel (e.g. `uname -a`):  Linux openpai-worker-01 4.4.0-186-generic #216-Ubuntu SMP Wed Jul 1 05:34:05 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- Hardware (e.g. core number, memory size, storage size, GPU type etc.):    
- - Intel (R) Xeon (R) Gold 5218 CPU @2.30GHz, 64 CPU(s), 2 threads per core
--  8 GPUs (nvidia tesla V100 with 32G memory)
-- 251 G memory
- Others:

**Anything else we need to know**:
This happens every time when I using multi-GPU mode with pytorch.

**Comment by Binyang2014** [comment]
Can you provide your job config. How many resources you request in job config? Seems you need to increase memory size in job config

**Comment by deepmeng** [comment]
@Binyang2014 
There is a resource instance from the job config.
`    resourcePerInstance:
      gpu: 3
      cpu: 27
      memoryMB: 120000`

**Comment by Binyang2014** [comment]
Can you increase your memory request to a large number such as `240000` and to see if this issue still happens?
And you can check the `Job Metrics` page to see if memory usage is hit the limitation.

PAI limits the resource occupied by each task though container. If you hit the limitation, OS will kill your job automatically

**Comment by deepmeng** [comment]
@Binyang2014 
Would you like to explain to me, in PyTorch's Dataloader function, the different parameter settings for `pin_memory` and `num_workers` will bring what different resource consumption in docker mode and non-docker mode?

**Comment by deepmeng** [comment]
@Binyang2014 
I can't add more memory. My guess is that this question involves the `pinned memory or page locked memory` and so on.

**Comment by Binyang2014** [comment]
Refer to this blog: https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/, if you use pin memory, you will copy data into pin memory first, it needs to alloc additional memory space for data.
And for pytorch `num_workers` will create multi python processes (not thread). So the common data structure can not be shared, it will also consume more resource.

In no non-docker mode, you can use all resources in your host. But in docker mode, OS will monitor your resource consumption. If memory usage exceed the limit, it will kill your process.  (Refer: https://www.kernel.org/doc/gorman/html/understand/understand016.html)

**Comment by deepmeng** [comment]
@Binyang2014 
I validated this guess myself three weeks ago. I think your detailed explanation is more convincing. Thank you for taking the time to answer my questions carefully. I have carefully read your explanation and the material you have provided. Thank you very much for your effort and responsible attitude.
